File: src/converter/plugins/AccessorsPlugin.ts
import ts, {GetAccessorDeclaration, SetAccessorDeclaration} from "typescript";
import {ConverterPlugin} from "../plugin.js";
import {ConverterContext} from "../context.js";
import {ifPresent, Render} from "../render.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {GeneratedFile} from "../generated.js";
import {InheritanceModifierService, inheritanceModifierServiceKey} from "./InheritanceModifierPlugin.js";

interface AccessorInfo {
    getter: GetAccessorDeclaration | undefined
    setter: SetAccessorDeclaration | undefined
}

const emptyAccessorInfo: AccessorInfo = {
    getter: undefined,
    setter: undefined,
}

export class AccessorsPlugin implements ConverterPlugin {
    private accessors = new Map<ts.Symbol, AccessorInfo>()
    private coveredAccessors = new Set<ts.Symbol>()

    setup(context: ConverterContext): void {
    }

    traverse(node: ts.Node, context: ConverterContext): void {
        if (ts.isGetAccessor(node) || ts.isSetAccessor(node)) {
            const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
            const typeChecker = typeScriptService?.program.getTypeChecker()

            const symbol = typeChecker?.getSymbolAtLocation(node.name)
            if (!symbol) return

            const accessorInfo = this.accessors.get(symbol) ?? {...emptyAccessorInfo}

            if (!this.accessors.has(symbol)) {
                this.accessors.set(symbol, accessorInfo)
            }

            if (ts.isGetAccessor(node)) {
                accessorInfo.getter = node
            }

            if (ts.isSetAccessor(node)) {
                accessorInfo.setter = node
            }
        }
    }

    render(node: ts.Node, context: ConverterContext, next: Render): string | null {
        if (ts.isSetAccessor(node) || ts.isGetAccessor(node)) {
            const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
            const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
            const inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)

            const inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)

            const typeChecker = typeScriptService?.program.getTypeChecker()

            const symbol = typeChecker?.getSymbolAtLocation(node.name)
            if (!symbol) return null

            if (this.coveredAccessors.has(symbol)) {
                checkCoverageService?.deepCover(node)
                return ""
            }

            checkCoverageService?.cover(node)

            if (ts.isSetAccessor(node)) {
                checkCoverageService?.deepCover(node.parameters[0])
            }

            this.coveredAccessors.add(symbol)

            const accessorInfo = this.accessors.get(symbol) ?? {...emptyAccessorInfo}

            const modifier = accessorInfo.setter !== undefined
                ? "var "
                : "val "

            const name = next(node.name)

            let type: string

            if (accessorInfo.getter && accessorInfo.getter.type) {
                type = next(accessorInfo.getter.type)
            } else if (accessorInfo.setter && accessorInfo.setter.parameters[0].type) {
                type = next(accessorInfo.setter.parameters[0].type)
            } else {
                type = "Any? /* type isn't declared */"
            }

            return `
${ifPresent(inheritanceModifier, it => `${it} `)}${modifier}${name}: ${type}
            `;
        }

        return null
    }

    generate(): GeneratedFile[] {
        return [];
    }
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/AccessorsPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.*
import io.github.sgrishchenko.karakum.util.escapeIdentifier
import js.objects.JsPlainObject
import typescript.*

@JsPlainObject
private external interface AccessorInfo {
    val getter: GetAccessorDeclaration?
    val setter: SetAccessorDeclaration?
}

private fun emptyAccessorInfo() = AccessorInfo(
    getter = null,
    setter = null,
)

class AccessorsPlugin : Plugin<Node> {
    private val accessors = mutableMapOf<Symbol, AccessorInfo>()
    private val coveredAccessors = mutableSetOf<Symbol>()

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) {
        if (isGetAccessor(node) || isSetAccessor(node)) {
            val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
            val typeChecker = typeScriptService?.program?.getTypeChecker()

            val name = when {
                isGetAccessor(node) -> node.name
                isSetAccessor(node) -> node.name
                else -> return
            }

            val symbol = typeChecker?.getSymbolAtLocation(name)
            if (symbol == null) return

            var accessorInfo = accessors[symbol] ?: emptyAccessorInfo()

            if (isGetAccessor(node)) {
                accessorInfo = accessorInfo.copy(getter = node)
            }

            if (isSetAccessor(node)) {
                accessorInfo = accessorInfo.copy(setter = node)
            }

            accessors[symbol] = accessorInfo
        }
    }

    override fun render(node: Node, context: Context, next: Render<Node>): String? {
        if (isSetAccessor(node) || isGetAccessor(node)) {
            val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
            val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
            val inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)

            val inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)

            val typeChecker = typeScriptService?.program?.getTypeChecker()

            val accessorName = when {
                isGetAccessor(node) -> node.name
                isSetAccessor(node) -> node.name
                else -> return null
            }

            val symbol = typeChecker?.getSymbolAtLocation(accessorName)
            if (symbol == null) return null

            if (symbol in this.coveredAccessors) {
                checkCoverageService?.deepCover(node)
                return ""
            }

            checkCoverageService?.cover(node)

            if (isSetAccessor(node)) {
                checkCoverageService?.deepCover(node.parameters.asArray()[0])
            }

            this.coveredAccessors.add(symbol)

            val accessorInfo = this.accessors[symbol] ?: emptyAccessorInfo()

            val modifier = if (accessorInfo.setter != null) "var " else "val "

            val name = escapeIdentifier(next(accessorName))
            val annotation = createKebabAnnotation(accessorName)

            val getterType = accessorInfo.getter?.type
            val setterType = accessorInfo.setter?.parameters?.asArray()?.get(0)?.type

            val type = if (getterType != null) {
                next(getterType)
            } else if (setterType != null) {
                next(setterType)
            } else {
                "Any? /* type isn't declared */"
            }

            return """
${ifPresent(annotation) { "${it}\n" }}${ifPresent(inheritanceModifier) { "$it "}}${modifier}${name}: $type
            """.trim();
        }

        return null
    }

    override fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()
}
